### ISSUE TITLE
Varying number of saves completed depending on browser

### ISSUE THREAD

**Issue by Svetlana154** [umple/umple#1747]
Depending on what browser and operating system is used, there is a different amount of saves occurring on the website, thus producing a skewed count of the edits made by the user. 

On my machine (Windows 10 with an Ubuntu Linux Subsystem), the saves on Chrome, Opera, and Edge behave similarly, while Firefox behaves completely differently. I added a console.log message for every function in umple_action.js that utilized umple_history.js's save method, and discovered the following:

<table>
<tr>
<th><em>Action</em></th>
<th><em>Saves in Chrome/Opera/Edge</em></th>
<th><em>Saves in FireFox</em></th>
</tr>
<tr>
<th>Upon load</th>
<th>loadFileCallback<br/>processTyping</th>
<th>loadFileCallback<br/>processTyping</th>
</tr>
<tr>
<th>Chose example</th>
<th>loadExampleCallback</th>
<th>loadExampleCallback<br/>textCallback</th>
</tr>
<tr>
<th>Edit code/text</th>
<th>processTyping</th>
<th>processTyping<br/>textCallback</th>
</tr>
<tr>
<th>Edit model</th>
<th>textCallback<br/>textCallback</th>
<th>textCallback (x4)</th>
</tr>
</table>

The "Action" column dictates what steps I took as the user and the "Saves" columns describes the purpose for which the save function was called (a new line indicates the save function was called multiple times).
As depicted by the table, save() is called way too many times for the textCallback purpose. 

This specific function is used numerous times in umple_action_diagram.js for scenarios when the users adds/removes elements from the model; however, upon further analysis, at most only one of the textCallbacks from each "Saves" column is related to a call from umple_action_diagram.js. 
As the user, both my text and model edits added a new attribute to a class. The newAttribute function from umple_action_diagram.js is responsible for the first textCallback save() for both categories of browsers - the rest of them require further research to figure out where they come from.

Note: Correcting this issue may directly affect Issue #1695 and Issue #1741

**Comment by TimLethbridge** [comment]
If this is to be addressed it needs to be addressed extremely cautiously. The reason for that is that over many years we got rid of very subtle bugs related to Ajax and the callbacks and the diagram and text getting updated correctly. Unfortunately this required some 'hacking' to get around browser differences, timing issues, etc. Also the people working on it varied in ability. This has resulted in complex, delicate code.  It seems to be working (notwithstanding issue #1741 where the example URL is erased). There is a high risk that certain types of change (e.g. on calls to key functions) might break something else, so i changes are made then many hours of testing on different browsers and with different examples (large, small) and different diagram types) would be needed by multiple people, comparing to the installed version, before we can release a change.

I am also not seeing evidence of repeated / redundant calls to the backend. This should be checked as it is a possible serious consequence of the symptoms expressed here. 

I suggest not making any changes without analysing and explaining the proposed changes here first.

One key test is whether undo and redo are working correctly at all times. If an undo undoes an example load, and a diagram edit and a text edit, and two undos undo two of these correctly, and a redo and two redos restore correctly, then that is an important test set. The table above hints that this may not in fact be working, but it seems OK for me on all Mac browsers (although I haven't tested deeply ... it was tested last year and was working well).

It is possible that the repeated saves are actually not 'really' happening (to backend and to undo queue) and that the calls to the functions are extraneous but not doing anything 'bad' in terms of unacceptable CPU usage. It may also be that it is only Firefox on Windows, or perhaps only one version of Firefox on Windows that is doing this.

All counts of edits related to the survey and hence to issue #1695 should be triggered by calls to history.save (umple_history.js:119) I would have thought. This function implements the queue for undoing and redoing, which seems to be working. But I haven't looked in detail at this.

Another issue is that callback functions from Ajax can be called multiple times, once per 'state' of the Ajax transaction. I haven't looked at the code, but this could perhaps be something to do with what is going on.

Maybe enhance the table in the main comment by giving the actual function name (with line number in the file) so it can be reliably found by others, and also indicate key arguments. E.g. processTyping is at line 2195. Its targets can be diagramEdit, umpleModelEditorText, codeMirrorEditor

**Comment by Svetlana154** [comment]
<h2>Additional information</h2>

loadFileCallback: 
- found on line 479 in umpleonline/scripts/umple_action.js

loadExampleCallback:
- found on line 1612 in umpleonline/scripts/umple_action.js

TextCallback:
- In function updateUmpleText
- found on line 1500 in umpleonline/scripts/umple_action.js

The files responsible for the saving function are umple_action.js, umple_action_diagram.js, and umple_system.js

**Comment by Svetlana154** [comment]
<h2>Proposed Changes</h2>
1. Duplicate TextCallbacks <br/><br/>
It seems like the function responsible for saving the Umple code for the textCallback purpose calls itself again after the .save method. Consider the function below (ignore all the console.log() that were used for the debugging process)
<br/>

```
Action.updateUmpleTextCallback = function(response)
{
  console.log("saved for TextCallback");
  TabControl.getCurrentHistory().save(response.responseText, "TextCallback");
  Action.freshLoad = true;
  
  Page.setUmpleCode(response.responseText, Action.update.codeChange);
  // DEBUG
  // Page.setFeedbackMessage("update text callback -");
  // Page.catFeedbackMessage(response.responseText);
  // Page.catFeedbackMessage("-");
  
  Page.hideLoading();

  if (DiagramEdit.textChangeQueue.length == 0) 
  {
    DiagramEdit.pendingChanges = false;
  }
  else{
    console.log('in else');
    DiagramEdit.doTextUpdate();
  }
  
  //Uncomment for testing purposes only - to update the image after updating the text
  //Action.updateUmpleDiagram();
} 
```
<br/>
The DiagramEdit.doTextUpdate() called in the else statement can be found on line 37 in umple_action_diagram.js and it does the following: <br/>

```
DiagramEdit.doTextUpdate = function()
{
  update = DiagramEdit.textChangeQueue.shift();
  Action.update = update;

  if(update.codeChange) {
    Page.hideGeneratedCode();
  }

  Action.ajax(Action.updateUmpleTextCallback,update.actionCode);
}
```
As evident by the ajax function, this would run the first function again, resulting in another TextCallback save.

-----------------------------
2. Increasing efficiency
<br/>
The following function is very commonly used after an edit in order to update various aspects in the back-end and it is found on line 16 in umple_action_diagram.js
<br/>

```
DiagramEdit.updateUmpleText = function(update)
{
  //console.log(update);
  //console.log("length: "+ DiagramEdit.textChangeQueue.length+"; pending changes: "+DiagramEdit.pendingChanges);
  if(DiagramEdit.textChangeQueue.length == 0 && !DiagramEdit.pendingChanges)
  {
    //console.log("entered if");
    DiagramEdit.pendingChanges = true;
    DiagramEdit.textChangeQueue.push(update);
    DiagramEdit.doTextUpdate();
  }
  else
  {
    //console.log("entered else");
    if (update.codeChange)
    DiagramEdit.textChangeQueue.push(update);
  }
  
  Action.umpleTypingActivity("diagramEdit");
}
```
The addition of an if statement such as if (update.codeChange) funnels out the updates that did not affect the code before adding them to the textChangeQueue, keeping the queue itself more organized.

-----------------------------
3. Reducing the TextCallback saves after loading an example or editing the code
<br/>
There are various functions in play for this issue, thus I wish to provide a breakdown of the entire process. Let's consider the loadExampleCallback save. <br/>

- The main function in charge of this is the loadExampleCallback, which sets the code according to the option chosen (dictated by the function that calls it - Action.loadExample on line 1554 in umple_action.js), saves the change in history, and updates the Umple diagram in response.

```
Action.loadExampleCallback = function(response)
{
  Action.freshLoad = true;
  Page.setUmpleCode(response.responseText);
  Page.hideLoading();
  console.log("saved for loadExample");
  TabControl.getCurrentHistory().save(response.responseText, "loadExampleCallback");
  Action.updateUmpleDiagram();
  Action.setCaretPosition("0");
  Action.updateLineNumberDisplay();
}
```
- The function of question there is Action.updateUmpleDiagram() which simply calls Action.updateUmpleDiagramForce(true) bringing us to line 2283 on umple_action.js which doesn't do much in terms of saving data except for calling an ajax function with Action.updateUmpleDiagramCallback().
- This function (found on line 2300 in umple_action.js) synchronizes the selected class diagram with the text. If we focus on the EditableClassDiagram, the following code will be executed:<br/>

```
    // Display editable class diagram
    if(Page.useEditableClassDiagram) {
      var newSystem = Json.toObject(diagramCode);
      UmpleSystem.merge(newSystem);
      UmpleSystem.update(); 
      
      //Apply readonly styles
      if (Page.readOnly) 
      {
        jQuery("span.editable").addClass("uneditable");
      }
    }
```

- It merges the new system/model into the old one and updates it using the UmpleSystem methods located in umple_system.js. If we take a closer look at UmpleSystem.update(), it calls UmpleSystem.trimOverlappingAssociations() for every class which eventually does:

```
DiagramEdit.updateUmpleText({
          actionCode: format("action=editAssociation&actionCode={0}",editAssociation),
          codeChange: false
        });
```

- This brings us back to umple_action_diagram.js and eventually the methods Action.updateUmpleTextCallback described in the first proposed change, thus saving the code once again for the TextCallback purpose (hence the bug).
<br/>
My proposition is to save the code only once under the "loadExampleCallback" purpose. This could be done by manipulating the order that the saves are being done in (updating the model first and then saving in the Action.loadExampleCallback function) and a bool variable to ensure that a TextCallback save doesn't occur when an example is loaded by a user.
<br/>The processTyping issue works in a similar fashion, so it could also be easily fixed in the same way.

----------------------
None of the proposed changes should change the redo and undo functions, but further testing is required to confirm this.

**Comment by Svetlana154** [comment]
<h2>Proposed Changes</h2>
It seems the program is running the TextCallback save after every model diagram edit. Hence, any modification made to it by selecting an example and/or altering the code causes "additional" saves. This could be fixed by introducing a bool variable, that would dictate when the actual model diagram has (not) been edited and thus, control when the TextCallback saves need to occur. 
<br/><br/>The scenarios, where TextCallback saves occurs multiple times, could also be resolved with the addition of that extra variable as well as the pre-existing indicators of diagram edits (specifically the DiagramEdit.textChangeQueue.length variable). All that is required is a proper modification of the if statement in umple_action.js's Action.updateUmpleTextCallback function, where the save for the TextCallback occurs. Placing the if statement any earlier in the operation disrupts specific elements of code that are needed for specific scenarios (e.g restricting the access to the top navigator bar or disabling the undo/redo options).

--------------------------------------
Different examples seems to run in different ways. I have noted that the "Accidents" example behaves uniquely from the rest of the examples in the Class Diagram list. Extra caution should be taken to make sure every example works the same after the changes to the saves.

**Comment by TimLethbridge** [comment]
This was fixed by #1766
